Refactor some ComponentsConvergence setup

joshuahansel opened the pull request:
- Created generic Convergence class `MultiPostprocessorConvergence` (in framework)
- Added some convergence-related methods to `Component`
- Refactored convergence setup in `FlowChannel1Phase` to take advantage of the above

After this PR,

- https://github.com/idaholab/moose/pull/32011

Refs #31038

joshuahansel committed:
Added common Convergence methods to Component

Refs #31038

joshuahansel committed:
Refactored FlowChannel1Phase convergence

Refs #31038

joshuahansel committed:
Created MultiPostprocessorConvergence

Refs #31038

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3447947/), step Docs: sync website on 9263e5b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/32084/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3447984/), step Generate coverage on 9263e5b wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3aea24</th>
      <th colspan=4>#32084 9263e5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.98%</td>
      <td>85.98%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>125766</td>
      <td>125822</td>
      <td>+56</td>
      <td>61</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>20505</td>
      <td>20510</td>
      <td>+5</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/32084/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/32084/coverage/framework)


## Modules coverage

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>3aea24</th>
      <th colspan=4>#32084 9263e5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.96%</td>
      <td>88.91%</td>
      <td>-0.04%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>15498</td>
      <td>15430</td>
      <td>-68</td>
      <td>78</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1924</td>
      <td>1924</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/32084/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/32084/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/32084/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on framework/src/convergence/MultiPostprocessorConvergence.C:
```diff
@@ -0,0 +1,109 @@
+unsigned int
+MultiPostprocessorConvergence::getMaxDescriptionLength() const
+{
+  size_t max_desc_length = 0;
```
```suggestion
  std::size_t max_desc_length = 0;
```

lindsayad commented on modules/thermal_hydraulics/include/components/Component.h:
```diff
@@ -457,6 +457,44 @@ class Component : public THMObject,
+   *
+   * @param[in] functor_name  Functor for which to compute step
+   * @param[in] property  Name of new step functor material property
+   * @param[in] functor_is_ad  Is the functor for which to compute step AD?
```
```suggestion
   * @param[in] functor_is_ad  Is the functor for which to compute the step AD?
```

lindsayad commented on modules/thermal_hydraulics/include/components/Component.h:
```diff
@@ -457,6 +457,44 @@ class Component : public THMObject,
+   * @param[in] descriptions  Description of each Postprocessor
+   * @param[in] tolerances  Tolerance for each check
+   */
+  void addStandardNonlinearConvergence(const std::vector<PostprocessorName> & postprocessors,
```
```suggestion
  void addNonlinearConvergence(const std::vector<PostprocessorName> & postprocessors,
```
"Standard" concerns me because this is a different "standard" convergence than that in the framework

joshuahansel replied:
`addNonlinearConvergence` sounds like it's the only way to add a nonlinear Convergence object for a component. Maybe just `addMultiPostprocessorConvergence`?

lindsayad commented on modules/thermal_hydraulics/src/components/Component.C:
```diff
@@ -246,6 +246,54 @@ Component::setSubdomainInfo(SubdomainID subdomain_id,
+void
+Component::addMaximumFunctorPostprocessor(const std::string & functor_name,
+                                          const std::string & pp_name,
+                                          Real normalization,
```
```suggestion
                                          const Real normalization,
```

lindsayad reviewed (commented):
Pretty cool seeing this combination of functor and convergence use. The framework test you added is also pretty awesome

joshuahansel committed:
Addressed review

- Wording improvements
- Minor type improvements
- Renamed addStandardNonlinearConvergence
- Allowed MultiPostprocessorConvergence to omit descriptions

Refs #31038

joshuahansel commented:
I hate when git suggestions cause format check to fail!

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3470169/), step Results summary on 173f4b3 wanted to post the following:

## Framework test summary

Compared against fb28c06 in job [civet.inl.gov/job/3464837](https://civet.inl.gov/job/3464837).

### Added tests

| Test                                               | Time (s)   | Memory (MB)   |
|----------------------------------------------------|------------|---------------|
| `convergence/multi_postprocessor_convergence.test` | 1.37       | 145.40        |
## Modules test summary

Compared against fb28c06 in job [civet.inl.gov/job/3464837](https://civet.inl.gov/job/3464837).

No change

lindsayad reviewed (approved):
I approve, assuming you won't merge it until the tests are passing haha
